File: clients/cpp/model/Mine.cpp
#include "Mine.hpp"

Mine::Mine() { }
Mine::Mine(int playerId, Vec2Double position, Vec2Double size, MineState state, std::shared_ptr<double> timer, double triggerRadius, ExplosionParams explosionParams) : playerId(playerId), position(position), size(size), state(state), timer(timer), triggerRadius(triggerRadius), explosionParams(explosionParams) { }
Mine Mine::readFrom(InputStream& stream) {
    Mine result;
    result.playerId = stream.readInt();
    result.position = Vec2Double::readFrom(stream);
    result.size = Vec2Double::readFrom(stream);
    switch (stream.readInt()) {
    case 0:
        result.state = MineState::PREPARING;
        break;
    case 1:
        result.state = MineState::IDLE;
        break;
    case 2:
        result.state = MineState::TRIGGERED;
        break;
    case 3:
        result.state = MineState::EXPLODED;
        break;
    default:
        throw std::runtime_error("Unexpected discriminant value");
    }
    if (stream.readBool()) {
        result.timer = std::shared_ptr<double>(new double());
        *result.timer = stream.readDouble();
    } else {
        result.timer = std::shared_ptr<double>();
    }
    result.triggerRadius = stream.readDouble();
    result.explosionParams = ExplosionParams::readFrom(stream);
    return result;
}
void Mine::writeTo(OutputStream& stream) const {
    stream.write(playerId);
    position.writeTo(stream);
    size.writeTo(stream);
    stream.write((int)(state));
    if (timer) {
        stream.write(false);
    } else {
        stream.write(true);
        stream.write((*timer));
    }
    stream.write(triggerRadius);
    explosionParams.writeTo(stream);
}
std::string Mine::toString() const {
    return std::string("Mine") + "(" +
        std::to_string(playerId) +
        position.toString() +
        size.toString() +
        "TODO" + 
        "TODO" + 
        std::to_string(triggerRadius) +
        explosionParams.toString() +
        ")";
}

File: clients/go/model/mine.go
package model

import (
	"io"

	mStream "../stream"
)

type Mine struct {
	PlayerId        int32
	Position        *Vec2Float64
	Size            *Vec2Float64
	State           MineState
	Timer           float64
	TriggerRadius   float64
	ExplosionParams *ExplosionParams
}

func NewMine(playerId int32, position *Vec2Float64, size *Vec2Float64, state MineState, timer float64,
	triggerRadius float64, explosionParams *ExplosionParams) *Mine {
	return &Mine{
		PlayerId:        playerId,
		Position:        position,
		Size:            size,
		State:           state,
		Timer:           timer,
		TriggerRadius:   triggerRadius,
		ExplosionParams: explosionParams,
	}
}
func ReadMine(reader io.Reader) *Mine {
	result := &Mine{
		PlayerId:        mStream.ReadInt32(reader),
		Position:        ReadVec2Float64(reader),
		Size:            ReadVec2Float64(reader),
		State:           ReadMineState(reader),
		TriggerRadius:   mStream.ReadFloat64(reader),
		ExplosionParams: ReadExplosionParams(reader),
	}

	if mStream.ReadBool(reader) {
		TimerValue := mStream.ReadFloat64(reader)
		result.Timer = TimerValue
	} else {
		result.Timer = -1
	}
	return result
}
func (value Mine) Write(writer io.Writer) {
	mStream.WriteInt32(writer, value.PlayerId)
	value.Position.Write(writer)
	value.Size.Write(writer)
	mStream.WriteInt32(writer, int32(value.State))
	if value.Timer < 0 {
		mStream.WriteBool(writer, false)
	} else {
		mStream.WriteBool(writer, true)
		mStream.WriteFloat64(writer, value.Timer)
	}
	mStream.WriteFloat64(writer, value.TriggerRadius)
	value.ExplosionParams.Write(writer)
}
